syntax = "proto3";

package com.alpha.backend.grpc;

option java_multiple_files = true;
option java_package = "com.alpha.backend.infrastructure.grpc.proto";
option java_outer_classname = "EmbeddingStreamProto";

// 데이터 스트리밍 서비스
service EmbeddingStreamService {
  // [기존] 서버 스트리밍 RPC: Batch Server가 요청하면, Python 서버가 Embedding 데이터를 스트림으로 전송
  rpc StreamEmbedding(StreamEmbeddingRequest) returns (stream RowChunk);

  // [신규] 클라이언트 스트리밍 RPC: Python 서버가 데이터를 스트림으로 Batch Server에 전송
  rpc IngestDataStream(stream IngestDataRequest) returns (IngestDataResponse);
}

// 스트리밍 요청
message StreamEmbeddingRequest {
  // 도메인 구분 (recruit, candidate, skill_dic)
  string domain = 1;
  // 마지막으로 처리된 UUID (체크포인트)
  // 이 값이 비어있으면 처음부터 전송
  optional string last_processed_uuid = 2;
  // 요청할 청크 크기 (0이면 서버 기본값 사용)
  optional int32 chunk_size = 3;
}

// ============================================================================
// 도메인별 데이터 청크 (oneof로 타입 구분)
// ============================================================================

// 데이터 청크 (도메인별 분기)
message RowChunk {
  oneof chunk_data {
    RecruitRowChunk recruit = 1;
    CandidateRowChunk candidate = 2;
    SkillEmbeddingDicRowChunk skill_dic = 3;
  }
}

// Recruit 청크
message RecruitRowChunk {
  repeated RecruitRow rows = 1;
}

// Candidate 청크
message CandidateRowChunk {
  repeated CandidateRow rows = 1;
}

// SkillEmbeddingDic 청크
message SkillEmbeddingDicRowChunk {
  repeated SkillEmbeddingDicRow rows = 1;
}

// ============================================================================
// 도메인별 Row 정의
// ============================================================================

// Recruit 도메인 (채용 공고) - v2
message RecruitRow {
  // recruit 테이블
  string id = 1;                    // UUID v7 (recruit_id)
  string position = 2;              // 포지션 원문 (NOT NULL)
  string company_name = 3;          // 회사명 (NOT NULL)
  int32 experience_years = 4;       // 요구 경력 (연, Nullable)
  string primary_keyword = 5;       // 주요 키워드 (Nullable)
  string english_level = 6;         // 영어 레벨 (Nullable)
  string published_at = 7;          // 게시 날짜 (ISO 8601, Nullable)

  // recruit_skill 테이블 (1:N)
  repeated string skills = 8;       // 요구 기술 스택 배열 (e.g., ["Java", "Spring", "AWS"])

  // recruit_description 테이블
  string long_description = 9;      // 채용 공고 원문 (Markdown, Nullable)
  string description_lang = 10;     // 원문 언어 (e.g., "en", "ko", Nullable)

  // recruit_skills_embedding 테이블
  repeated float skills_vector = 11; // 기술 스택 벡터 (384d)
}

// Candidate 도메인 (후보자) - Flat DTO 구조
message CandidateRow {
  string candidate_id = 1;          // UUID v7
  string position_category = 2;     // 직종 카테고리
  int32 experience_years = 3;       // 경력 (연)
  string original_resume = 4;       // 원문 이력서
  repeated string skills = 5;       // 보유 스킬 배열 (e.g., ["Java", "Python"])
  repeated float skills_vector = 6; // 기술 스택 벡터 (768d)
}

// SkillEmbeddingDic 도메인 (기술 사전)
message SkillEmbeddingDicRow {
  string skill = 1;                 // 스킬명 (PK)
  string position_category = 2;     // 직종 카테고리
  repeated float skill_vector = 3;  // 스킬 벡터 (768d)
}

// --- 신규 메시지 정의 ---

// 클라이언트 스트리밍 요청 메시지
message IngestDataRequest {
  // 첫 번째 요청에만 메타데이터를 포함
  oneof request_type {
    IngestMetadata metadata = 1;
    // 실제 데이터 청크 (JSON 문자열 형태)
    bytes data_chunk = 2;
  }
}

// 스트리밍 메타데이터
message IngestMetadata {
  string domain = 1;      // "recruit", "candidate" 등
  string file_name = 2;
  int32 vector_dimension = 3; // 벡터 차원
}

// 클라이언트 스트리밍 응답 메시지
message IngestDataResponse {
  bool success = 1;
  int32 received_chunks = 2; // 수신된 총 청크 수
  string message = 3;
}
